collated/main/Kowalski985.md

```diff
@@ -0,0 +1,516 @@
+            ArrayList<String> missingPrefixes = parser.getMissingPrefixes(arguments);
+            if (lastCharIsStartOfPrefix(commandBoxText)) {
+                state = AutocompleteState.COMMAND_COMPLETE_PREFIX;
+                return;
+            }
+
+            if (lastTwoCharactersArePrefix(commandBoxText)) {
+                setIndexToOneIfNeeded();
+                if (missingPrefixes.size() > possibleAutocompleteResults.size()) {
+                    possibleAutocompleteResults = missingPrefixes;
+                }
+                state = AutocompleteState.COMMAND_CYCLE_PREFIX;
+                return;
+            }
+
+            possibleAutocompleteResults = missingPrefixes;
+            state = AutocompleteState.COMMAND_NEXT_PREFIX;
+        } else {
+            state = AutocompleteState.COMMAND;
+        }
+
+    }
+
+    private boolean isCyclingThroughCommands(String commandWord) {
+        return state.equals(AutocompleteState.MULTIPLE_COMMAND) && textInCommandBox.length() == commandWord.length();
+    }
+
+    /**
+     * Returns true if the the autocompleter was not cycling through possible options
+     * in it's previous state
+     */
+    private void resetIndexIfNeeded() {
+        if (!state.equals(AutocompleteState.MULTIPLE_COMMAND)) {
+            resultIndex = 0;
+        }
+    }
+
+    private void setIndexToOneIfNeeded() {
+        if (!state.equals(AutocompleteState.COMMAND_CYCLE_PREFIX)) {
+            resultIndex = 1;
+        }
+    }
+
+    private void resetCountingAndMaxIndexIfNeeded() {
+        if (!state.equals(AutocompleteState.INDEX)) {
+            countingIndex = 1;
+        }
+    }
+
+    private boolean needIndex(String arguments) {
+        return  (state.equals(AutocompleteState.INDEX) && lastCharIsDigit(textInCommandBox))
+                || !(containsIndex(arguments));
+    }
+
+    /**
+     * Check if the index field in the {@code String} has already been entered
+     */
+    private boolean containsIndex(String arguments) {
+        String parameters = arguments;
+        Prefix[] prefixes = AutocompleteCommand.ALL_PREFIXES;
+        if (lastTwoCharactersArePrefix(arguments)) {
+            parameters = arguments + SPACE;
+        }
+        ArgumentMultimap argMap = ArgumentTokenizer.tokenize(parameters, prefixes);
+        String index = argMap.getPreamble();
+        return (!index.equals(EMPTY_STRING) && isNumeric(index));
+    }
+
+    /**
+     * Returns true if the string is numeric
+     */
+    private boolean isNumeric (String index) {
+        try {
+            return index.matches("[0-9]+");
+        } catch (NullPointerException e) {
+            return false;
+        }
+    }
+
+    /**
+     * Returns true if the last character of the {@code String} is a digit
+     */
+    private boolean lastCharIsDigit(String text) {
+        if (text.length() < 1) {
+            return false;
+        }
+        return Character.isDigit(text.charAt(text.length() - 1));
+    }
+
+
+
```

Comment:
Use blank lines sparingly. It's good to have blank lines to separate logical code blocks, but too many blank lines will increase the height of the code, which increases the amount of scrolling.